File: src/radar/diff.py
"""Unified-diff parsing — one pass over the PR diff shared by conformance and
(later) capture (FR-ARCH-1)."""

from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path

_FILE_HEADER_RE = re.compile(r"^diff --git a/(?P<a>\S+) b/(?P<b>\S+)", re.MULTILINE)


@dataclass
class FileDiff:
    path: str  # post-change path
    text: str  # the file's full diff segment (headers + hunks)


def parse_unified_diff(text: str) -> list[FileDiff]:
    matches = list(_FILE_HEADER_RE.finditer(text))
    diffs: list[FileDiff] = []
    for i, m in enumerate(matches):
        end = matches[i + 1].start() if i + 1 < len(matches) else len(text)
        segment = text[m.start():end]
        path = m.group("b")
        if "+++ /dev/null" in segment:  # deleted file — keep pre-change path
            path = m.group("a")
        diffs.append(FileDiff(path=path, text=segment.rstrip("\n")))
    return diffs


def load_diff(path: str | Path) -> list[FileDiff]:
    return parse_unified_diff(Path(path).read_text())

File: src/diff.ts
/**
 * Unified-diff parsing — one pass over the PR diff shared by conformance and
 * (later) capture (FR-ARCH-1).
 */
import { readFileSync } from "node:fs";

export interface FileDiff {
  path: string; // post-change path
  text: string; // the file's full diff segment (headers + hunks)
}

const FILE_HEADER_RE = /^diff --git a\/(?<a>\S+) b\/(?<b>\S+)/gm;

export function parseUnifiedDiff(text: string): FileDiff[] {
  const matches = [...text.matchAll(FILE_HEADER_RE)];
  const diffs: FileDiff[] = [];
  for (let i = 0; i < matches.length; i++) {
    const m = matches[i];
    const start = m.index ?? 0;
    const end = i + 1 < matches.length ? (matches[i + 1].index ?? text.length) : text.length;
    const segment = text.slice(start, end);
    let path = m.groups!.b;
    if (segment.includes("+++ /dev/null")) {
      path = m.groups!.a; // deleted file — keep pre-change path
    }
    diffs.push({ path, text: segment.replace(/\n+$/, "") });
  }
  return diffs;
}

export function loadDiff(path: string): FileDiff[] {
  return parseUnifiedDiff(readFileSync(path, "utf8"));
}
